src/engine/controls/bpmcontrol.cpp

```diff
@@ -750,118 +798,115 @@ double BpmControl::getBeatMatchPosition(
-                &dThisBeatLength,
+        getBeatContext(pBeats,
+                thisPosition,
+                &thisPrevBeat,
+                &thisNextBeat,
+                &dThisBeatLengthFrames,
                 nullptr);
         // now we either have a useful next beat or there is none
-        if (dThisNextBeat == -1) {
+        if (thisNextBeat == mixxx::kInvalidFramePos) {
             // We can't match the next beat, give up.
-            return dThisPosition;
+            return thisPosition;
         }
     } else {
         if (kLogger.traceEnabled()) {
             kLogger.trace() << "BpmControl::getBeatMatchPosition up to date"
-                            << dThisPrevBeat << dThisPosition << dThisNextBeat;
+                            << thisPrevBeat << thisPosition << thisNextBeat;
         }
         // We are between the previous and next beats so we can try a standard
         // lookup of the beat length.
-        getBeatContextNoLookup(
-                dThisPosition,
-                dThisPrevBeat,
-                dThisNextBeat,
-                &dThisBeatLength,
+        getBeatContextNoLookup(thisPosition,
+                thisPrevBeat,
+                thisNextBeat,
+                &dThisBeatLengthFrames,
                 nullptr);
     }
 
     TrackPointer otherTrack = pOtherEngineBuffer->getLoadedTrack();
-    mixxx::BeatsPointer otherBeats = otherTrack ? otherTrack->getBeats() : mixxx::BeatsPointer();
+    mixxx::BeatsPointer otherBeats =
+            otherTrack ? otherTrack->getBeats() : mixxx::BeatsPointer();
 
     // If either track does not have beats, then we can't adjust the phase.
     if (!otherBeats) {
-        return dThisPosition;
+        return thisPosition;
     }
 
-    double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
+    mixxx::FramePos otherPosition = samplePosToFramePos(pOtherEngineBuffer->getExactPlayPos());
 
-    double dOtherPrevBeat = -1;
-    double dOtherNextBeat = -1;
-    double dOtherBeatLength = -1;
+    mixxx::FramePos otherPrevBeat = mixxx::kInvalidFramePos;
+    mixxx::FramePos otherNextBeat = mixxx::kInvalidFramePos;
+    mixxx::FrameDiff_t dOtherBeatLength = -1;
     double dOtherBeatFraction = -1;
-    if (!BpmControl::getBeatContext(
-                otherBeats,
-                dOtherPosition,
-                &dOtherPrevBeat,
-                &dOtherNextBeat,
+    if (!BpmControl::getBeatContext(otherBeats,
+                otherPosition,
+                &otherPrevBeat,
+                &otherNextBeat,
                 &dOtherBeatLength,
                 &dOtherBeatFraction)) {
-        return dThisPosition;
+        return thisPosition;
     }
 
     if (dOtherBeatLength == -1 || dOtherBeatFraction == -1) {
         // the other Track has no usable beat info, do not seek.
-        return dThisPosition;
+        return thisPosition;
     }
 
-    double dThisSampleRate = m_pBeats->getSampleRate();
+    double dThisSampleRate = m_pTrack->getSampleRate();
```

Comment:
This breaks the atomic assumption of m_pBeats.
The whole code works, because it holds a atomic pointer to a consistent beats dataset.
The track is a new dataset which changes independently.
A calculations must be done based into the same atomic pointer. 
We must assure that this pointer (m_pBeats) is only touched at one point during the callback.